File: examples/tutorials/rest-api-annotated-service/src/main/java/example/armeria/server/blog/BadRequestExceptionHandler.java
package example.armeria.server.blog;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import com.linecorp.armeria.common.HttpRequest;
import com.linecorp.armeria.common.HttpResponse;
import com.linecorp.armeria.common.HttpStatus;
import com.linecorp.armeria.server.ServiceRequestContext;
import com.linecorp.armeria.server.annotation.ExceptionHandlerFunction;

/**
 * Handles an {@link IllegalArgumentException} and returns an {@link HttpResponse} with
 * {@link HttpStatus#BAD_REQUEST}.
 */
public class BadRequestExceptionHandler implements ExceptionHandlerFunction {

    private static final ObjectMapper mapper = new ObjectMapper();

    @Override
    public HttpResponse handleException(ServiceRequestContext ctx, HttpRequest req, Throwable cause) {
        if (cause instanceof IllegalArgumentException) {
            final String message = cause.getMessage();
            final ObjectNode objectNode = mapper.createObjectNode();
            objectNode.put("error", message);
            return HttpResponse.ofJson(HttpStatus.BAD_REQUEST, objectNode);
        }
        return ExceptionHandlerFunction.fallthrough();
    }
}

File: examples/annotated-http-service-scala/src/main/scala/server/blog/BadRequestExceptionHandler.scala
package server.blog

import com.fasterxml.jackson.databind.ObjectMapper
import com.linecorp.armeria.common.{HttpRequest, HttpResponse, HttpStatus}
import com.linecorp.armeria.server.ServiceRequestContext
import com.linecorp.armeria.server.annotation.ExceptionHandlerFunction

class BadRequestExceptionHandler extends ExceptionHandlerFunction :
  val mapper = ObjectMapper()

  override def handleException(ctx: ServiceRequestContext, req: HttpRequest, cause: Throwable): HttpResponse =
    if cause.isInstanceOf[IllegalArgumentException] then
      val message = cause.getMessage
      val objectNode = mapper.createObjectNode()
      objectNode.put("error", message)
      HttpResponse.ofJson(HttpStatus.BAD_REQUEST, objectNode)
    else
      ExceptionHandlerFunction.fallthrough()

